### ISSUE TITLE
SWIG misses apply_simple_mixing_rule when building Java wrapper

### ISSUE THREAD

**Issue by pkirkham** [CoolProp/CoolProp#2274]
### Description

I'm attempting to follow through the examples ([http://www.coolprop.org/coolprop/Cubics.html#adding-your-own-fluids](http://www.coolprop.org/coolprop/Cubics.html#adding-your-own-fluids)) for adding a custom fluid and then obtaining mixture properties. I'm using a Java wrapper for CoolProp. When adding a custom fluid it seems that this is an essential step to set the dummy interaction parameters with apply_simple_mixing_rule. Unfortunately, the method apply_simple_mixing_rule appears to be missing from CoolProp.Java generated by SWIG. Without it, PropsSI will throw an error that it could not match the binary pair. I presume that this is because no binary pair data has been set between the custom fluid and other fluids?

I thought the issue might be related to my installation of cmake, SWIG etc. as I've compiled the CoolProp code myself in order to assign a specific package (org.coolprop) to the generated code. Yet when I look back at the pre-compiled Java wrappers from 6.4.1 and earlier, this method is also missing. It appears that for some reason SWIG is not including apply_simple_mixing_rule in the JNI stubs.

### Steps to Reproduce

1. Code to add custom fluid in Java:
    ```java
   String fluidstring = "[\n  {\n    \"CAS\": \"000-0-00\",\n    \"Tc\": 400.0,\n    \"Tc_units\": \"K\",\n"
                + "    \"acentric\": 0.1,\n    \"aliases\": [\n    ],\n    \"molemass\": 0.04,\n"
                + "    \"molemass_units\": \"kg/mol\",\n    \"name\": \"FAKEFLUID\",\n    \"pc\": 4000000.0,\n"
                + "    \"pc_units\": \"Pa\"\n  }\n]";
   CoolProp.add_fluids_as_JSON("PR", fluidstring);
   System.out.println(CoolProp.PropsSI("T", "P", 101325, "Q", 0.0, "SRK::FAKEFLUID")+ " K");
    ```
   This step works as expected and produces the text output "247.0950826605273 K"
2. Method CoolProp.apply_simple_mixing_rule does not exist, so try skipping and just do a mixture calculation using our new custom fluid.
   ```java
    System.out.println(CoolProp.PropsSI("T", "P", 101325, "Q", 0.0, "HEOS::FAKEFLUID-SRK[0.3]&Ethane-SRK[0.7]"));
    ```
   This throws an error.

**Expected behavior:** Ideally the apply_simple_mixing_rule method would be accessible via Java using SWIG. Otherwise, in the event that the interaction parameters cannot be set, defaulting to a value of 1.0 for all interaction pairs between an existing fluid and a custom fluid might be sensible?

**Actual behavior:** Throws an error: java.lang.RuntimeException: Initialize failed for backend: "HEOS", fluid: "FAKEFLUID-SRK&Ethane-SRK" fractions "[ 0.3000000000, 0.7000000000 ]"; error: Could not match the binary pair [000-0-00,74-84-0] - for now this is an error.

### Versions

**CoolProp Version:** 6.5.0dev. 
**Operating System and Version:** Windows 11
**Access Method:** Java wrapper compiled using cmake 3.26.4 and swigwin-4.1.1. 

### Additional Information

This is my first experience with SWIG and C++, so I'm not sure where the best place to look is. I noticed that in /src/pybind11_interface.cxx there is an include at the top of the file for "Backends/Helmholtz/MixtureParameters.h". This is not included in CoolProp.i, so I tried adding it there, and to SWIG_DEPENDENCIES in CMakeLists.txt. After rebuilding and recompiling, neither change made apply_simple_mixing_rule appear.

I'm wondering if it's because the header files listed in CoolProp.i are all located in the /include/ folder, and that I should therefore be using a path like "../src/Backends/Helmholtz/MixtureParameters.h"?

**Comment by ibell** [comment]
Yes you need to be looking in ``src/CoolProp.i`` for which headers get included in the SWIG interface (ignore the pybind11 stuff, that is for python binding), and I think you are right that you want relative paths, but relative to what is always a tricky question

**Comment by ibell** [comment]
Or you might want to add a small passthrough function analogously to how ``add_fluids_as_JSON`` is handled in ``CoolProp.h`` and ``CoolProp.cpp``

**Comment by pkirkham** [comment]
Thanks for the guidance Ian. My reference to the python bindings was an observation that the MixtureParameters.h file, which includes the apply_simple_mixing_rule function, appeared in the python bindings, but not in the Java bindings. In any event, I tried using relative paths without success.

The brute force approach you suggested worked. I created a small function `add_dummy_mixing_rule` in `CoolProp.h` and `CoolProp.cpp`  which simply calls `apply_simple_mixing_rule` with the "Lorentz-Berthelot" rule. This time SWIG picked up the method and exposed it through the Java wrapper.

In C++ (CoolProp.cpp):
```cpp
void apply_dummy_mixing_rule(const std::string& identifier1, const std::string& identifier2) {
    apply_simple_mixing_rule(identifier1, identifier2, "Lorentz-Berthelot");
}
```

Generated by SWIG in Java (CoolProp.java):
```java
public static void apply_dummy_mixing_rule(String identifier1, String identifier2) {
    CoolPropJNI.apply_dummy_mixing_rule(identifier1, identifier2);
}
```

Now we can set dummy binary interaction parameters between our custom fluid and Ethane for step [2] of the code above before doing the mixture calculation using our custom fluid:
```java
String cas_ff = "000-0-00";
String cas_c2 = CoolProp.get_fluid_param_string("Ethane", "CAS");
CoolProp.set_config_bool(configuration_keys.OVERWRITE_BINARY_INTERACTION, true);
CoolProp.apply_dummy_mixing_rule(cas_ff, cas_c2);
System.out.println(CoolProp.PropsSI("T", "P", 101325, "Q", 0.0, "HEOS::FAKEFLUID-SRK[0.3]&Ethane-SRK[0.7]"));
```

This succeeds and gives the output "190.93809486697978" as per the custom fluid example (http://www.coolprop.org/coolprop/Cubics.html#adding-your-own-fluids).

Whilst it still isn't clear why SWIG doesn't pick up the function as-is when generating the wrapper, at least there is a practical workaround.

**Comment by ibell** [comment]
A better solution would be to mirror the arguments, not changing the function name, and allow the kind of simple mixing rule to be passed through to the call. If you did a PR with this change I would be happy to review it.

**Comment by pkirkham** [comment]
Thanks for the suggestion. I've tried to implement so that the function name and arguments are not changed. The reason for not doing so in the first place was that it refused to compile. I tried adding a new namespace, but in the end the only thing that would keep the compiler happy was putting a new `add_simple_mixing_rule` in `CoolProp.cpp` and renaming in `MixtureParameters.cpp`. It's not elegant, but it compiles and the function is now wrapped by SWIG.

I created a PR with the change that works.

I should also add that this is my first ever PR for an open source project. I'm not a developer by trade -- just an advanced hobbyist. That said, I am a petroleum engineer and I've been writing my own tools using Java for over 20 years. This has included writing my own implementation, from scratch, for the Predictive Peng Robinson 78 cubic EoS. Whilst I've ported some code from Fortran, C/C++ into Java in the past, I've never written in the language or contributed to an open source project. I guess I've just been waiting for the right project to come along. I'm very impressed with what everyone has done in creating CoolProp and thank you all for your efforts.

**Comment by ibell** [comment]
As I mentioned in the PR, I think it is a bad idea to change the core of CoolProp. I prefer we figure out why the simple passthrough wasn't working in the first place. 

Glad to have you. You might also be interested in: https://github.com/usnistgov/teqp if you are into EOS implementations

**Comment by ibell** [comment]
Closing as part of an issue-tracker cleanup — this issue has had no activity in over two years.

*To be transparent: this comment and the close action are being performed by Claude (Anthropic's AI coding assistant), running on behalf of @ibell. Ian reviewed and approved the bulk-close criteria; I am the one executing the action.*

If you would like this issue reconsidered, please feel free to **re-open it** — a fresh comment with current behaviour against the latest CoolProp release is the most useful way to revive the discussion.

— Claude (acting for @ibell)
